src/spritecache.cpp

```diff
@@ -669,203 +649,133 @@ bool LoadNextSprite(int load_index, SpriteFile &file, uint file_sprite_id)
-		if (s->size & S_FREE_MASK) {
-			MemBlock *next = NextBlock(s);
-			MemBlock temp;
-			SpriteID i;
-
-			/* Since free blocks are automatically coalesced, this should hold true. */
-			assert(!(next->size & S_FREE_MASK));
-
-			/* If the next block is the sentinel block, we can safely return */
-			if (next->size == 0) break;
-
-			/* Locate the sprite belonging to the next pointer. */
-			for (i = 0; GetSpriteCache(i)->ptr != next->data; i++) {
-				assert(i != _spritecache_items);
-			}
+	int bpp = BlitterFactory::GetCurrentBlitter()->GetScreenDepth();
+	uint target_size = (bpp > 0 ? _sprite_cache_size * bpp / 8 : 1) * 1024 * 1024;
+	if (_spritecache_bytes_used > target_size) {
+		DeleteEntriesFromSpriteCache(_spritecache_bytes_used - target_size + 512 * 1024);
+	}
 
-			GetSpriteCache(i)->ptr = s->data; // Adjust sprite array entry
-			/* Swap this and the next block */
-			temp = *s;
-			memmove(s, next, next->size);
-			s = NextBlock(s);
-			*s = temp;
+	if (_sprite_lru_counter >= 0xC0000000) {
+		Debug(sprite, 3, "Fixing lru {}, inuse={}", _sprite_lru_counter, _spritecache_bytes_used);
 
-			/* Coalesce free blocks */
-			while (NextBlock(s)->size & S_FREE_MASK) {
-				s->size += NextBlock(s)->size & ~S_FREE_MASK;
+		for (SpriteCache &sc : _spritecache) {
+			if (!sc.data.empty()) {
+				if (sc.lru > 0x80000000) {
+					sc.lru -= 0x80000000;
+				} else {
+					sc.lru = 0;
+				}
 			}
-		} else {
-			s = NextBlock(s);
 		}
+		_sprite_lru_counter -= 0x80000000;
 	}
 }
 
-/**
- * Delete a single entry from the sprite cache.
- * @param item Entry to delete.
- */
-static void DeleteEntryFromSpriteCache(uint item)
+void *AllocSprite(size_t mem_req)
 {
-	/* Mark the block as free (the block must be in use) */
-	MemBlock *s = (MemBlock*)GetSpriteCache(item)->ptr - 1;
-	assert(!(s->size & S_FREE_MASK));
-	s->size |= S_FREE_MASK;
-	GetSpriteCache(item)->ptr = nullptr;
-
-	/* And coalesce adjacent free blocks */
-	for (s = _spritecache_ptr; s->size != 0; s = NextBlock(s)) {
-		if (s->size & S_FREE_MASK) {
-			while (NextBlock(s)->size & S_FREE_MASK) {
-				s->size += NextBlock(s)->size & ~S_FREE_MASK;
-			}
-		}
-	}
+	assert(_last_sprite_allocation.empty());
+	_last_sprite_allocation.resize(mem_req);
+	return _last_sprite_allocation.data();
 }
 
-static void DeleteEntryFromSpriteCache()
+void SpriteCache::AssignLastAllocSprite(void *ptr)
 {
-	uint best = UINT_MAX;
-	int cur_lru;
-
-	Debug(sprite, 3, "DeleteEntryFromSpriteCache, inuse={}", GetSpriteCacheUsage());
-
-	cur_lru = 0xffff;
-	for (SpriteID i = 0; i != _spritecache_items; i++) {
-		SpriteCache *sc = GetSpriteCache(i);
-		if (sc->type != SpriteType::Recolour && sc->ptr != nullptr && sc->lru < cur_lru) {
-			cur_lru = sc->lru;
-			best = i;
-		}
+	_spritecache_bytes_used -= this->data.size();
+	if (ptr != nullptr) {
+		assert(static_cast<void *>(_last_sprite_allocation.data()) == ptr);
+		this->data = std::move(_last_sprite_allocation);
+		_spritecache_bytes_used += this->data.size();
```

Comment:
vector has a swap() method which might be useful rather than messing about with std::move (which technically makes the moved object invalid, I believe.)

Reply:
std::move does not make the moved object invalid.
It isn't required to do a move in the other direction as well.

Reply:
Okay, most moved objects are in a "valid but unspecified state", but vector is specifically valid and empty, apparently.